// apify-cli/dist/lib/outputs.js
import chalk from 'chalk';
function internalLog(options) {
    if (options.stdoutOutput) {
        console.log(...options.stdoutOutput);
    }
    if (options.stderrOutput) {
        console.error(...options.stderrOutput);
    }
}
export function simpleLog(options) {
    internalLog({
        [options.stdout ? 'stdoutOutput' : 'stderrOutput']: [options.message],
    });
}
export function error(options) {
    internalLog({
        [options.stdout ? 'stdoutOutput' : 'stderrOutput']: [chalk.red('Error:'), options.message],
    });
}
export function warning(options) {
    internalLog({
        [options.stdout ? 'stdoutOutput' : 'stderrOutput']: [chalk.rgb(254, 90, 29).bold('Warning:'), options.message],
    });
}
export function success(options) {
    internalLog({
        [options.stdout ? 'stdoutOutput' : 'stderrOutput']: [chalk.green('Success:'), options.message],
    });
}
export function run(options) {
    internalLog({
        [options.stdout ? 'stdoutOutput' : 'stderrOutput']: [chalk.gray('Run:'), options.message],
    });
}
export function info(options) {
    internalLog({
        [options.stdout ? 'stdoutOutput' : 'stderrOutput']: [chalk.white('Info:'), options.message],
    });
}
export function link(options) {
    internalLog({
        [options.stdout ? 'stdoutOutput' : 'stderrOutput']: [chalk.blue(options.message), options.url],
    });
}
